### PR TITLE
[Bug] [PLC] [Builtins] Unlifting to 'Int' can fail

### PR DESCRIPTION
**Pull request by effectfully**
This fixes an embarrassing bug found by @bezirg. 

Pre-submit checklist:
- Branch
    - [ ] Commit sequence broadly makes sense
    - [ ] Key commits have useful messages
    - [ ] Relevant tickets are mentioned in commit messages
- PR
    - [ ] Self-reviewed the diff
    - [ ] Useful pull request description
    - [ ] Reviewer requested
- If you updated any cabal files or added Haskell packages:
    - [ ] `$(nix-build default.nix -A pkgsLocal.updateMaterialized)` to update the materialized Nix files
   - [ ] Update `hie-*.yaml` files if needed
- If you changed any Haskell files:
   - [ ] `$(nix-shell shell.nix --run fix-stylish-haskell)` to fix any formatting issues
- If you changed any Purescript files:
   - [ ] `$(nix-shell shell.nix --run fix-purty)` to fix any formatting issues

Pre-merge checklist:
- [ ] Someone approved it
- [ ] Commits have useful messages
- [ ] Review clarifications made it into the code
- [ ] History is moderately tidy; or going to squash-merge

### CODE REVIEW

File: plutus-core/src/Language/PlutusCore/Constant/Typed.hs [hunk 1]
```diff
@@ -340,3 +340,36 @@ instance KnownBuiltinType term String        => KnownType term String
 instance KnownBuiltinType term Char          => KnownType term Char
 instance KnownBuiltinType term ()            => KnownType term ()
 instance KnownBuiltinType term Bool          => KnownType term Bool
+
+{- Note [Int as Integer]
+We represent 'Int' as 'Integer' in PLC and check that an 'Integer' fits into 'Int' when
+unlifting constants (and fail if it doesn't).
```
Comment (michaelpj) [inline plutus-core/src/Language/PlutusCore/Constant/Typed.hs:346]: This seems wrong to me. I thought that rather:
- Our `Integer` builtin type *really is* `Integer`: it's arbitrary-precision. We don't want to deal with overflow issues most of the time!
- *Some* of our builtins don't handle `Integer`, they really need `Int`. Mostly the bytestring ones. So we should do a restricting conversion `Integer -> Int` at the point where we take the argument, and the builtin should fail (as builtins normally fail) if that conversion doesn't work.

Comment (michaelpj) [reply plutus-core/src/Language/PlutusCore/Constant/Typed.hs:346]: In other words, the previous behaviour seemed right, we just needed to handle the case where the restriction doesn't work in the semantics of the builtin function.

Comment (effectfully) [reply plutus-core/src/Language/PlutusCore/Constant/Typed.hs:346]: Err, stupid docs, sorry. This sentence should be

> We represent 'Int' as 'Integer' in PLC and check that an 'Integer' fits into 'Int' when
unlifting constants of type 'Int' (and fail if it doesn't).

(added the extra "of type 'Int').

If you look at the code, what happens there is exactly what you described:

> So we should do a restricting conversion `Integer -> Int` at the point where we take the argument, and the builtin should fail (as builtins normally fail) if that conversion doesn't work.

Comment (effectfully) [reply plutus-core/src/Language/PlutusCore/Constant/Typed.hs:346]: `Integer` is still a separate type and lifting from / unlifting to `Integer` works as before. I only added an instance for `Int` that does the checking.

Comment (michaelpj) [reply plutus-core/src/Language/PlutusCore/Constant/Typed.hs:346]: But we're still doing the failure "outside" the builtin semantics, right? Which is why you have this whole problem with the errors. What I'm suggesting is we write

```
specialTake :: Integer -> ByteString -> ByteString
specialTake i bs | tooSmall i = ... -- whatever you do to indicate that builtins fail, I can't remember
specialTake i bs = take (fromIntegral i) bs
```

Comment (michaelpj) [reply plutus-core/src/Language/PlutusCore/Constant/Typed.hs:346]: So then the builtin "really" takes an `Integer`, so we have no need to pretend that we can unlift `Int`s.

Comment (effectfully) [reply plutus-core/src/Language/PlutusCore/Constant/Typed.hs:346]: Yeah, we can do that, but I kinda like to pretend we can unlift `Int`s as that's what we really do in the end. Ok, I agree broken semantics is not worth it, I'll try implementing the `AsEvaluationFailure` thing, it should improve other parts as well, so maybe that'll convince you.

File: plutus-core/src/Language/PlutusCore/Evaluation/Result.hs [hunk 2]
```diff
@@ -4,19 +4,34 @@
...
     ) where
 
 import           Language.PlutusCore.Pretty
 
 import           Control.Applicative
+import           Control.Lens
+import           Control.Monad.Except
 import           PlutusPrelude
 
+class AsEvaluationFailure err where
```
Comment (michaelpj) [inline plutus-core/src/Language/PlutusCore/Evaluation/Result.hs:25]: No `makeClassyPrisms`? Worth a note why not.

File: plutus-core/src/Language/PlutusCore/Evaluation/Result.hs [hunk 3]
```diff
@@ -4,19 +4,34 @@
...
+    _EvaluationFailure :: Prism' err ()
+
+instance AsEvaluationFailure () where
+    _EvaluationFailure = id
+
+_EvaluationFailureVia :: Eq err => err -> Prism' err ()
```
Comment (michaelpj) [inline plutus-core/src/Language/PlutusCore/Evaluation/Result.hs:31]: This is a little mysterious on first reading...

**Review by michaelpj — APPROVED** [review]

File: plutus-core/src/Language/PlutusCore/Constant/Typed.hs [hunk 4]
```diff
@@ -340,3 +351,36 @@ instance KnownBuiltinType term String        => KnownType term String
...
+
+There's a catch, though: an out-of-bounds error is not an internal one -- it's a normal
+evaluation failure, but unlifting errors have this connotation of being "internal".
+We could make 'readKnown' return an @m (EvaluationResult a)@, but that makes it more awkward
+to use (in particular, in functions like 'readKnownCek').
+
+Another option is be to define and use @AsEvaluationFailure@ in addition to 'AsUnliftingError'.
```
Comment (michaelpj) [inline plutus-core/src/Language/PlutusCore/Constant/Typed.hs:364]: You did this in the end, right? so the comment is out of date

Comment (effectfully) [reply plutus-core/src/Language/PlutusCore/Constant/Typed.hs:364]: Yes, the docs are not updated (I did not want to update them before you approve the solution). Will do.

**Comment by effectfully** [comment]
Comments addressed.

**Review by michaelpj — APPROVED** [review]

**Comment by bezirg** [comment]
I reviewed it, it looks fine to me although I don't know much about the innards of makeKnown/readKnown; I initially though that the change would be like @michaelpj 's suggestion of just replacing `fromIntegral` with some check+failure. 

I have a question, shouldn't <https://github.com/input-output-hk/plutus/blob/5759480a9d343cc0f8ab2318e2360fc4d104e65b/plutus-tx/src/Language/PlutusTx/Builtins.hs#L72  >
this be changed as well to reflect the new behaviour?

**Comment by effectfully** [comment]
> I initially though that the change would be like @michaelpj 's suggestion of just replacing `fromIntegral` with some check+failure.

I find the current way of doing things more ergonomic while amounting to the same check+failure machinery under the hood.

> be changed as well to reflect the new behaviour?

Great question. @michaelpj

**Comment by michaelpj** [comment]
That is a great question. I should make it clearer what the conditions on these functions are in the doc comments. They're supposed to be "off-chain equivalents" of the on-chain functions. In this case I think what's there is fine: it does `BS.take`, and fails with an error if the `Integer` is out of range.

**Comment by effectfully** [comment]
I have bad news for ya:

```
>>> fromIntegral ((2 :: Integer) ^ (100 :: Integer)) :: Int
0
```

`fromIntegral` eventually delegates to [`integerToInt`](https://hackage.haskell.org/package/integer-gmp-1.0.3.0/docs/src/GHC.Integer.Type.html#integerToInt), which eventually delegates to `bigNatToInt` implemented as

```haskell
bigNatToInt :: BigNat -> Int#
bigNatToInt (BN# ba#) = indexIntArray# ba# 0#
```

i.e. `bigNatToInt` simply truncates the array without checking anything.

**Pull request merged**